drivers: i2s: Add support for nRF I2S peripherals

anangl opened the pull request:
Add a shim that allows using the nrfx I2S driver via the Zephyr API.
Add also missing devicetree nodes representing the I2S peripherals
in the nRF52 Series SoCs.

Bases on #34250.

Resolves #11702.

anangl committed:
drivers: i2s: Add support for nRF I2S peripherals

Add a shim that allows using the nrfx I2S driver via the Zephyr API.
Add also missing devicetree nodes representing the I2S peripherals
in the nRF52 Series SoCs.
Extend the "nordic,nrf-i2s" binding with a new property that allows
specifying the clock source to be used by the I2S peripheral (so that
it is possible to use HFXO for better accurracy of the peripheral clock
or, in the nRF53 Series SoCs, to use the dedicated audio oscillator).

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

galak reviewed (dismissed):
Are there any boards that i2s can be enabled on so we can test the new driver?

galak commented:
> Are there any boards that i2s can be enabled on so we can test the new driver?

Said another way, please make sure there that at least one test exists that builds this driver. (ie tests/drivers/i2s/i2s_api)

parthitce commented on drivers/i2s/i2s_nrfx.c:
```diff
@@ -0,0 +1,788 @@
+#include <logging/log.h>
+LOG_MODULE_REGISTER(i2s_nrfx, CONFIG_I2S_LOG_LEVEL);
+
+struct queue {
```
Can be replaced with `struct k_msgq` instead of custom queue.

anangl replied:
Applied.

anangl replied:
@parthitce I addressed your comment. Would you mind taking a look again?

parthitce replied:
@anangl LGTM, thanks

mbolivar-nordic commented on drivers/i2s/i2s_nrfx.c:
```diff
@@ -0,0 +1,788 @@
+	static void data_handler##idx(nrfx_i2s_buffers_t const *p_released,  \
+				      uint32_t status)			     \
+	{								     \
+		data_handler(DEVICE_DT_INST_GET(idx), p_released, status);   \
```
I think you mean `DEVICE_DT_GET(I2S(idx))`, right?

anangl replied:
Yes. Fixed.

carlescufi commented:
@adamkondraciuk as the author of the original attempt at this, could you please take a look?

kshitizdange commented on drivers/i2s/i2s_nrfx.c:
```diff
@@ -0,0 +1,788 @@
+
+	if (i2s_cfg->frame_clk_freq == 0) { /* -> reset state */
+		purge_queue(drv_data, dir);
+		if (dir == I2S_DIR_TX || I2S_DIR_BOTH) {
+			drv_data->tx_configured = false;
+			memset(&drv_data->tx.nrfx_cfg, 0,
+			       sizeof(drv_data->tx.nrfx_cfg));
+			memset(&drv_data->tx.config, 0,
+			       sizeof(drv_data->tx.config));
+		}
+		if (dir == I2S_DIR_RX || I2S_DIR_BOTH) {
```
Shouldn't these be `if (dir == I2S_DIR_TX || dir == I2S_DIR_BOTH)`  and `if (dir == I2S_DIR_RX || dir == I2S_DIR_BOTH)`?

anangl replied:
Thanks for spotting that. Fixed.

davedesro commented:
@anangl . I'm excited to see this one merged. It's a long time coming! Will code review comments be reviewed soon?

anangl commented:
Rebased on updated #34250.

anangl commented:
- dropped the commit that switched from *-pin to *-gpios properties
- used 'k_msgq` instead of the custom queue implementation
- addressed the other review comments

anangl commented:
Rebased on master after #34250 is merged.

anangl committed:
samples: drivers: Add I2S echo sample

Add a sample intended to show how to use I2S drivers in processing of
audio streams. The sample starts both the RX and TX streams and adds
a simple echo effect to the received audio signal.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl commented:
Added a sample that shows the use of the newly introduced `I2S_DIR_BOTH` enumeration value and that ensures that the I2S nrfx driver shim is built by CI.

mbolivar-nordic commented on drivers/i2s/i2s_nrfx.c:
```diff
@@ -0,0 +1,708 @@
+			 &i2s_nrf_drv_api);
+
+/* Currently, only one I2S instance is supported. */
+I2S_NRFX_DEVICE(0);
```
Not blocking, but the above comment might be clearer as:

```
/* Existing SoCs only have one I2S instance */
```

That makes it clear it's not a driver limitation.

mbolivar-nordic commented on samples/drivers/i2s/echo/README.rst:
```diff
@@ -0,0 +1,40 @@
+********
+
+This sample demonstrates how to use an I2S driver in a simple processing of
+an audio stream. It configures and start both the RX and TX streams and then
```
```suggestion
an audio stream. It configures and starts both the RX and TX streams and then
```

mbolivar-nordic commented on samples/drivers/i2s/echo/README.rst:
```diff
@@ -0,0 +1,40 @@
+This sample demonstrates how to use an I2S driver in a simple processing of
+an audio stream. It configures and start both the RX and TX streams and then
+mixes the original signal with its delayed form that is buffered, providing
+this way a simple echo effect.
```
```suggestion
a simple echo effect.
```

mbolivar-nordic commented on samples/drivers/i2s/echo/README.rst:
```diff
@@ -0,0 +1,40 @@
+************
+
+The sample uses the WM8731 audio CODEC that can be found, for example,
+on the Audio Codec Shield, but it can be easily adapted to use other
```
Is there a reference for this shield?

anangl replied:
Added.

mbolivar-nordic commented on samples/drivers/i2s/echo/README.rst:
```diff
@@ -0,0 +1,40 @@
+streams.
+This sample has been tested on :ref:`nrf52840dk_nrf52840` (nrf52840dk_nrf52840)
+and :ref:`nrf5340dk_nrf5340` (nrf5340dk_nrf5340_cpuapp), and provides overlay
+files for both these boards.
```
```suggestion
files for both of these boards.
```

mbolivar-nordic commented on samples/drivers/i2s/echo/dts/bindings/wolfson,wm8731.yaml:
```diff
@@ -0,0 +1,8 @@
+# Copyright (c) 2021 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+description: WM8731 Audio CODEC
```
Why isn't this in the top level dts/bindings directory?

anangl replied:
It seemed more appropriate to me, as such nodes are used only in this one sample and I'm not providing a driver for this CODEC, only a simple initialization routine. But I can move the file if you think it would be better to place it in the top level bindings.

mbolivar-nordic replied:
>> Why isn't this in the top level dts/bindings directory?
>
> It seemed more appropriate to me, as such nodes are used only in this one sample and I'm not providing a driver for this CODEC, only a simple initialization routine.

OK, that's what I figured. That seems reasonable, but could you please add a comment explaining that in the file itself?

Also cc @galak

anangl replied:
Added.

mbolivar-nordic commented on samples/drivers/i2s/echo/src/main.c:
```diff
@@ -0,0 +1,304 @@
+
+#define SW0_NODE        DT_ALIAS(sw0)
+#if DT_NODE_HAS_STATUS(SW0_NODE, okay)
+#define SW0_GPIO_CTLR   DT_GPIO_CTLR(SW0_NODE, gpios)
```
Please use a `struct gpio_dt_spec`. I have an example in ` samples: button: devicetree cleanups ` in https://github.com/zephyrproject-rtos/zephyr/pull/34589 that relies on a new macro for potentially missing nodes, but this could be done even with a plain GPIO_DT_SPEC_GET and some extra ifdefs, and it would still be simpler.

anangl replied:
Applied.

mbolivar-nordic commented on samples/drivers/i2s/echo/src/main.c:
```diff
@@ -0,0 +1,304 @@
+		return false;
+	}
+
+	ret = gpio_pin_configure(sw0_dev, SW0_GPIO_PIN,
```
This would then become a call to gpio_pin_configure_dt(), which saves boilerplate.

mbolivar-nordic commented on samples/drivers/i2s/echo/src/main.c:
```diff
@@ -0,0 +1,304 @@
+		return false;
+	}
+
+	ret = gpio_pin_interrupt_configure(sw0_dev,
```
And this would be come a call to gpio_pin_interrupt_configure_dt().

mbolivar-nordic reviewed (changes requested):
Looks good! Some docs things and minor requests in the sample.

anangl committed:
twister: Add dt_nodelabel_enabled devicetree filter function

This function takes a node label (not a label property) and returns
true if a node with such label exists and the node is enabled.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

galak commented:
We really need some platform that stresses this driver with `tests/drivers/i2s/i2s_api`

anangl committed:
tests: drivers: i2s_api: Allow testing on nRF52840 DK and nRF5340 DK

Provide configurations for the nRF52840 DK and nRF5340 DK boards.
Adjust the test to cover specifics of I2S peripherals on nRF SoCs
(need of starting RX and TX simultaneously, lack of internal loopback).

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl commented:
@galak `tests/drivers/i2s/i2s_api` passes now for nRF52840 DK and nRF5340 DK (see below the output for nRF52840 DK). Apart from improving the driver, I also had to adjust quite a bit the test itself to cover specifics of the nRF I2S peripherals. I also fixed the MPU fault (Data Access Violation) that occurred in test cases run in user mode (I think @mmahadevan108 signaled the same problem in https://github.com/zephyrproject-rtos/zephyr/pull/32054#issue-568558271).
```
*** Booting Zephyr OS build zephyr-v2.5.0-3105-g941cd8849aca  ***
Running test suite i2s_loopback_test
===================================================================
START - test_i2s_tx_transfer_configure_0
 PASS - test_i2s_tx_transfer_configure_0 in 0.1 seconds
===================================================================
START - test_i2s_rx_transfer_configure_0
 PASS - test_i2s_rx_transfer_configure_0 in 0.1 seconds
===================================================================
START - test_i2s_transfer_short
1->OK
2->OK
Using data offset: 2
1<-OK
3->OK
2<-OK
3<-OK
 PASS - test_i2s_transfer_short in 0.15 seconds
===================================================================
START - test_i2s_transfer_long
 PASS - test_i2s_transfer_long in 0.407 seconds
===================================================================
START - test_i2s_rx_sync_start
RX sync start is not applicable with I2S_DIR_BOTH
 SKIP - test_i2s_rx_sync_start in 0.5 seconds
===================================================================
START - test_i2s_rx_empty_timeout
 PASS - test_i2s_rx_empty_timeout in 2.1 seconds
===================================================================
START - test_i2s_transfer_restart
1->OK
2->OK
1<-OK
Stop transmission
Start transmission
3->OK
2<-OK
3<-OK
 PASS - test_i2s_transfer_restart in 1.15 seconds
===================================================================
START - test_i2s_transfer_rx_overrun
E: Failed to allocate next RX buffer: -12
 PASS - test_i2s_transfer_rx_overrun in 0.409 seconds
===================================================================
START - test_i2s_transfer_tx_underrun
E: Next buffers not supplied on time
E: Cannot write in state: 4
 PASS - test_i2s_transfer_tx_underrun in 0.619 seconds
===================================================================
Test suite i2s_loopback_test succeeded
Running test suite i2s_states_test
===================================================================
START - test_i2s_tx_transfer_configure_1
 PASS - test_i2s_tx_transfer_configure_1 in 0.1 seconds
===================================================================
START - test_i2s_rx_transfer_configure_1
 PASS - test_i2s_rx_transfer_configure_1 in 0.1 seconds
===================================================================
START - test_i2s_state_not_ready_neg
E: Device is not configured
E: Device is not configured
E: Device is not configured
E: Device is not configured
E: Device is not configured
E: Device is not configured
E: Device is not configured
E: Device is not configured
E: Device is not configured
E: Device is not configured
E: Device is not configured
 PASS - test_i2s_state_not_ready_neg in 0.28 seconds
===================================================================
START - test_i2s_state_ready_neg
 PASS - test_i2s_state_ready_neg in 0.1 seconds
===================================================================
START - test_i2s_state_running_neg
 PASS - test_i2s_state_running_neg in 0.25 seconds
===================================================================
START - test_i2s_state_stopping_neg
 PASS - test_i2s_state_stopping_neg in 0.9 seconds
===================================================================
START - test_i2s_state_error_neg
E: Failed to allocate next RX buffer: -12
E: Cannot write in state: 4
 PASS - test_i2s_state_error_neg in 0.412 seconds
===================================================================
Test suite i2s_states_test succeeded
Running test suite i2s_user_loopback_test
===================================================================
START - test_i2s_tx_transfer_configure_0
 PASS - test_i2s_tx_transfer_configure_0 in 0.1 seconds
===================================================================
START - test_i2s_rx_transfer_configure_0
 PASS - test_i2s_rx_transfer_configure_0 in 0.1 seconds
===================================================================
START - test_i2s_transfer_short
1->OK
2->OK
1<-OK
3->OK
2<-OK
3<-OK
 PASS - test_i2s_transfer_short in 0.15 seconds
===================================================================
START - test_i2s_transfer_long
 PASS - test_i2s_transfer_long in 0.407 seconds
===================================================================
START - test_i2s_rx_sync_start
RX sync start is not applicable with I2S_DIR_BOTH
 SKIP - test_i2s_rx_sync_start in 0.5 seconds
===================================================================
START - test_i2s_rx_empty_timeout
 PASS - test_i2s_rx_empty_timeout in 2.1 seconds
===================================================================
START - test_i2s_transfer_restart
1->OK
2->OK
1<-OK
Stop transmission
Start transmission
3->OK
2<-OK
3<-OK
 PASS - test_i2s_transfer_restart in 1.15 seconds
===================================================================
START - test_i2s_transfer_tx_underrun
E: Next buffers not supplied on time
E: Cannot write in state: 4
 PASS - test_i2s_transfer_tx_underrun in 0.619 seconds
===================================================================
START - test_i2s_transfer_rx_overrun
E: Failed to allocate next RX buffer: -12
 PASS - test_i2s_transfer_rx_overrun in 0.409 seconds
===================================================================
Test suite i2s_user_loopback_test succeeded
Running test suite i2s_user_states_test
===================================================================
START - test_i2s_tx_transfer_configure_1
 PASS - test_i2s_tx_transfer_configure_1 in 0.1 seconds
===================================================================
START - test_i2s_rx_transfer_configure_1
 PASS - test_i2s_rx_transfer_configure_1 in 0.1 seconds
===================================================================
START - test_i2s_state_not_ready_neg
E: Device is not configured
E: Device is not configured
E: Device is not configured
E: Device is not configured
E: Device is not configured
E: Device is not configured
E: Device is not configured
E: Device is not configured
E: Device is not configured
E: Device is not configured
E: Device is not configured
 PASS - test_i2s_state_not_ready_neg in 0.29 seconds
===================================================================
START - test_i2s_state_ready_neg
 PASS - test_i2s_state_ready_neg in 0.1 seconds
===================================================================
START - test_i2s_state_running_neg
 PASS - test_i2s_state_running_neg in 0.25 seconds
===================================================================
START - test_i2s_state_stopping_neg
 PASS - test_i2s_state_stopping_neg in 0.9 seconds
===================================================================
START - test_i2s_state_error_neg
E: Failed to allocate next RX buffer: -12
E: Cannot write in state: 4
 PASS - test_i2s_state_error_neg in 0.412 seconds
===================================================================
Test suite i2s_user_states_test succeeded
===================================================================
PROJECT EXECUTION SUCCESSFUL
```

mmahadevan108 commented on tests/drivers/i2s/i2s_api/Kconfig:
```diff
@@ -12,3 +12,25 @@ config I2S_TEST_SEPARATE_DEVICES
+	  Use wiring between the data-out and data-in pins for looping back
+	  data. This option is intended to be used for devices that do not
+	  provide the internal loopback functionality.
+
```
This seems to be same as `config I2S_TEST_SEPARATE_DEVICES`

nikhil30081995 replied:
Hello how can i test the following Testcase. I also would like to test I2S peripheral for my nRF52840 dk board.

nikhil1983 commented:
Hello, I was trying to use above example using updated repo, but the I2S device doesn't get enabled in device tree itself even after configuration in proj.conf and overlay file. 


But when I myself add I2S peripheral configuration in .dtsi .dts file (in my old repo) it enables the device in zephyr device tree. 

Does anyone know what can be the issue here, as I really want to test this I2S peripheral using Zephyr API on nRF52804. 


Is it the issue of multiple zephyr old and new versions in my NCS folder ? I have tried replacing old repo with the new one in NCS folder, but it gives me some west error when I try to import project into segger.  

Any assistance would be very much helpful. 


Note: Old repo means (without nrfx_i2s drivers for nRF52804 based on Zephyr API)

sam131208 commented:
Hello, I was trying to use above example on nrf52833, but the I2S device is stop after running some minutes. I check and found, the function data_handler of i2s_nrfx.c can not free mem_lab in sometimes.

sam131208 commented:
The  variable next_tx_buffer_needed of i2s_nrfx.c must clear before trigger again。

anangl commented:
> The variable next_tx_buffer_needed of i2s_nrfx.c must clear before trigger again。

@sam131208 Thanks for pointing this out. Fixed.

anangl commented:
Rebased.

anangl commented:
Rebased.

anangl commented:
#36648 needs to go in first.

anangl commented:
Changes introduced by the [recent push](https://github.com/zephyrproject-rtos/zephyr/compare/15a5e073973e9a5e134b95a4a2efb08889f652d7..a70e4554142fb615de52a49a417a65559c93ede9):
- added the possibility to use the dedicated audio PLL clock (ACLK) as the clock source on nRF53 Series SoCs
- extended the functionality of Button 2 in the echo sample so that it is possible to restart streams after they are stopped; this revealed some imperfections in the added nRF I2S driver and those were also corrected

The i2s_api test still passes for both nRF52840 DK and nRF5340 DK.

anangl commented:
@galak Would you mind taking a look again?
@mnkp Could you check if my changes in the i2s_api test haven't broken anything for other platforms, like the Atmel SAM one?

mniestroj commented on drivers/i2s/i2s_nrfx.c:
```diff
@@ -0,0 +1,927 @@
+	COND_CODE_1(DT_NODE_HAS_PROP(I2S(idx), name##_pin),	\
+		    (DT_PROP(I2S(idx), name##_pin)),		\
+		    (NRFX_I2S_PIN_NOT_USED))
+#define I2S_CLK_SRC(idx)  DT_ENUM_IDX(I2S(idx), clock_source)
```
What about using `DT_ENUM_UPPER_TOKEN()` instead, so that we don't need to keep enum order in sync between DT bindings and this driver?

anangl replied:
Good point, thanks! Applied.

mnkp reviewed (commented):
Looking at the "tests: drivers: i2s_api: Allow testing on nRF52840 DK and nRF5340 DK" commit it feels like the changes required to support `I2S_DIR_BOTH` direction are quite intrusive and there are not so many common parts. I.e. some configuration code is shared between BOTH and RX/TX mode but the actual testcase is different. I wonder if it was not better to keep them separately in different compilation units, i.e. add new testcases like `test_i2s_dir_both_loopback.c`, `test_i2s_dir_both_states.c`. The testcases would only need to make sure that they are skipped if they require dir option not supported by the driver.

This would also make it a bit easier to test future drivers that support `I2S_DIR_BOTH` as well as `I2S_DIR_RX`, `I2S_DIR_TX`. And we would be testing that the driver returns `-ENOSYS` for the direction it doesn't support.

anangl commented:
> Looking at the "tests: drivers: i2s_api: Allow testing on nRF52840 DK and nRF5340 DK" commit it feels like the changes required to support `I2S_DIR_BOTH` direction are quite intrusive and there are not so many common parts. I.e. some configuration code is shared between BOTH and RX/TX mode but the actual testcase is different.

The testcases are functionally the same. The only difference is in triggering commands - for `CONFIG_I2S_TEST_USE_I2S_DIR_BOTH` by definition commands for TX and RX must be triggered simultaneously. But sending and receiving of data blocks stayed as it was.

> I wonder if it was not better to keep them separately in different compilation units, i.e. add new testcases like `test_i2s_dir_both_loopback.c`, `test_i2s_dir_both_states.c`. The testcases would only need to make sure that they are skipped if they require dir option not supported by the driver.

Yes, this would make the code look cleaner. I'll apply your suggestion. I'm not really happy with how the code looks right now, with such number of `if (IS_ENABLED())`.

anangl commented:
Rebased after #36648 has been merged.

anangl commented:
Refactored changes in the i2s_api test.
